README.txt
README for bouncestudio
=======================

== DESCRIPTION

The bouncestudio ruby extension wraps the Bounce Studio C API 
from Boogie Tools, Inc. This allows a simple method of determining
if a given raw email message is a hard bounce, an "over quota" 
bounce, etc. 

== SYNOPSIS

# Rubygems magic
require 'rubygems'
require 'bouncestudio'

# First, set your license. If you don't have a BounceStudio license 
# from BoogieTools, you can still test out the software. It'll just be
# too slow for production use. Go to http://www.boogietools.com/ to learn
# more. Using an environment variable is recommended to keep your license 
# code out of your source. 
BS_LICENSE = ENV["BS_LICENSE"] || ""

# Create the BounceStudio worker
bs = BoogieTools::BounceStudio.new(BS_LICENSE)

# Get your message, from a file, with Net::POP, etc
raw_message = "Subject: I love you"

# Issue the check. 
result = bs.check(raw_message)

# If you're not licensed, this is where you'll get a delay and 
# the following message:
#
#    Thank you for trying the BounceStudio API.
#    To register, please visit http://www.boogietools.com.


# Do whatever you want based on the following result codes.

# Type Description
# 0   NON BOUNCE
# 10  HARD BOUNCE
# 20  SOFT BOUNCE - General
# 21  SOFT BOUNCE - Dns Failure
# 22  SOFT BOUNCE - Mailbox Full
# 23  SOFT BOUNCE - Message Size Too Large
# 30  BOUNCE WITH NO EMAIL ADDRESS
# 40  GENERAL BOUNCE
# 50  MAIL BLOCK - General
# 51  MAIL BLOCK - Known Spammer
# 52  MAIL BLOCK - Spam Detected
# 53  MAIL BLOCK - Attachment Detected
# 54  MAIL BLOCK - Relay Denied
# 60  AUTO REPLY
# 70  TRANSIENT BOUNCE
# 80  SUBSCRIBE REQUEST
# 90  UNSUBSCRIBE REQUEST
# 100 CHALLENGE RESPONSE


== REQUIREMENTS

Besides Ruby and Rubygems, you'll of course need the BounceStudio
library. It is available for Linux or Windows. This wrapper has only
been tested with the Linux version. 

== INSTALL

Using Rubygems:

* sudo gem install bouncestudio

== LICENSE

(The MIT License)

Copyright (c) 2007 BoogieTools, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.




CHANGELOG.txt
2/27/07: Slight change to prevent calling bsBounceStudio_init() multiple times. Caused massive memory usage.
2/6/07: Initial upload to RubyForge


test/test_helper.rb
require 'test/unit'
require File.dirname(__FILE__) + '/../lib/bouncestudio'
if ENV['BS_LICENSE']
  BS_LICENSE = ENV['BS_LICENSE']
else
  puts "Set the environment variable BS_LICENSE with your BoogieTools license code to avoid tarpit"
  BS_LICENSE = "nolicense"
end



test/bouncestudio_test.rb
require File.dirname(__FILE__) + '/test_helper.rb'

class BouncestudioTest < Test::Unit::TestCase

  def setup
    license_code = BS_LICENSE
    @studio = BoogieTools::BounceStudio.new license_code
    @raw = IO.read File.join(File.dirname(__FILE__), "raw_message.txt")
  end
  
  def test_truth
    assert true
  end

  def test_studio_type
    assert_equal Object, BoogieTools::BounceStudio.superclass
    assert_equal BoogieTools::BounceStudio, @studio.class
  end

  def test_check
    bounce_code = @studio.check(@raw)
    assert bounce_code.is_a?(Fixnum)
    assert_equal bounce_code, 10
  end
  
  def test_body
    body = @studio.body(@raw)
    assert body.is_a?(String) 
  end

  def test_from_address
    from_address = @studio.from_address(@raw)
    assert from_address.is_a?(String)
    assert !from_address.empty?
    assert_equal from_address, "MAILER-DAEMON@mail.coinet.com".downcase
  end

  def test_from_friendly_name
    from_friendly_name = @studio.from_friendly_name(@raw)
    assert from_friendly_name.is_a?(String)
    assert from_friendly_name.empty?
  end

  def test_header
    header = @studio.header(@raw)
    assert header.is_a?(String)
    assert !header.empty?
  end

  def test_reply_to_address
    reply_to_address = @studio.reply_to_address(@raw)  
    assert reply_to_address.is_a?(String)
    assert reply_to_address.empty?
  end

  def test_reply_to_friendly_name
    reply_to_friendly_name = @studio.reply_to_friendly_name(@raw)  
    assert reply_to_friendly_name.is_a?(String)
    assert reply_to_friendly_name.empty?
  end

  def test_subject
    subject = @studio.subject(@raw)  
    assert subject.is_a?(String)
    assert !subject.empty?
    assert_equal subject.chomp, "failure notice"
  end

  def test_to_address
    to_address = @studio.to_address(@raw)  
    assert to_address.is_a?(String)
    assert !to_address.empty?
    assert_equal to_address, "joebob@joe.net"
  end

  def test_to_friendly_name
    to_friendly_name = @studio.to_friendly_name(@raw)  
    assert to_friendly_name.is_a?(String)
    assert to_friendly_name.empty?
  end

  def test_custom_header
    custom_header = @studio.custom_header(@raw, "X-OriginalArrivalTime".downcase)  
    assert custom_header.is_a?(String)
    assert custom_header.empty?

    date_header = @studio.custom_header(@raw, "Date".downcase)
    assert date_header.is_a?(String)
    assert !date_header.empty?
    assert_equal date_header.chomp, "21 Aug 2000 21:08:21 -0000"
  end

  def test_orig_custom_header
    orig_custom_header = @studio.orig_custom_header(@raw, "X-OriginalArrivalTime".downcase)
    assert orig_custom_header.is_a?(String)
    assert !orig_custom_header.empty?
    assert_equal orig_custom_header.chomp, "21 Aug 2000 21:08:20.0944 (UTC) FILETIME=[EF5F8500:01C00BB3]"
  end
end


ext/BounceStudio.h
/*
 * Copyright (c) 1999-2006 BoogieTools, Inc. All rights reserved
 *
 */

#ifndef BOUNCESTUDIO_H
#define BOUNCESTUDIO_H

/* constants returned by BSBOUNCECHECK */
#define BS_UNDETERMINED            0 /* Undetermined */
#define BS_HARD                   10 /* HARD BOUNCE */
#define BS_SOFT_GENERAL           20 /* SOFT BOUNCE - General */
#define BS_SOFT_DNSFAILURE        21 /* SOFT BOUNCE - Dns Failure */
#define BS_SOFT_MAILBOXFULL       22 /* SOFT BOUNCE - Mailbox Full */
#define BS_SOFT_MESSAGETOBIG      23 /* SOFT BOUNCE - Message Size Too Large */
#define BS_NOEMAIL                30 /* BOUNCE - email address could not be retrieved. */
#define BS_GENERAL                40 /* GENERAL BOUNCE */
#define BS_MAILBLOCK_GENERAL      50 /* MAIL BLOCK - General */
#define BS_MAILBLOCK_SPAMMER      51 /* MAIL BLOCK - Known Spammer */
#define BS_MAILBLOCK_SPAMDETECTED 52 /* MAIL BLOCK - Spam Content Detected */
#define BS_MAILBLOCK_ATTACHMENT   53 /* MAIL BLOCK - Attachment Detected */
#define BS_MAILBLOCK_RELAYDENIED  54 /* MAIL BLOCK - Relay Denied */
#define BS_AUTOREPLY              60 /* AUTO REPLY */
#define BS_TRANSIENT              70 /* TRANSIENT BOUNCE */
#define BS_SUBSCRIBE              80 /* SUBSCRIBE REQUEST */
#define BS_UNSUBSCRIBE            90 /* UNSUBSCRIBE REQUEST */
#define BS_CHALLENGERESPONSE     100 /* CHALLENGE-RESPONSE MESSAGE */

#ifdef __cplusplus
extern "C" {
#endif

/* Initialize The BounceStudio API Engine */
int bsBounceStudio_init();
int bsBounceCheck(char *sMessageText, char **sEmailAddress, 
                  char *sEmailIgnoreList, char *sKey);
void bsGetHeader(char *sMessageText, char **sReturn);
void bsGetBody(char *sMessageText, char **sReturn);
void bsGetSubject(char *sMessageText, char **sReturn);
void bsGetToAddress(char *sMessageText, char **sReturn);
void bsGetToFriendlyName(char *sMessageText, char **sReturn);
void bsGetReplyToAddress(char *sMessageText, char **sReturn);
void bsGetReplyToFriendlyName(char *sMessageText, char **sReturn);
void bsGetFromAddress(char *sMessageText, char **sReturn);
void bsGetFromFriendlyName(char *sMessageText, char **sReturn);
void bsGetCustomHeader(char *sMessageText, char **sReturn,
                       char *sCustomHeader);
void bsGetOrigCustomHeader(char *sMessageText, char **sReturn,
                           char *sCustomHeader);

#ifdef __cplusplus
}
#endif

#endif


ext/bounce_studio.c
#include "ruby.h"
#include "BounceStudio.h"

/*
Type Description
0   NON BOUNCE
10  HARD BOUNCE
20  SOFT BOUNCE - General
21  SOFT BOUNCE - Dns Failure
22  SOFT BOUNCE - Mailbox Full
23  SOFT BOUNCE - Message Size Too Large
30  BOUNCE WITH NO EMAIL ADDRESS
40  GENERAL BOUNCE
50  MAIL BLOCK - General
51  MAIL BLOCK - Known Spammer
52  MAIL BLOCK - Spam Detected
53  MAIL BLOCK - Attachment Detected
54  MAIL BLOCK - Relay Denied
60  AUTO REPLY
70  TRANSIENT BOUNCE
80  SUBSCRIBE REQUEST
90  UNSUBSCRIBE REQUEST
100 CHALLENGE RESPONSE
*/

/*
 * BounceStudio.new()
 */
static char *license; 
static int bs_initialized;

static VALUE rb_bs_init(VALUE self, VALUE lic_string) {
  if( bs_initialized != 1 ) {
    bsBounceStudio_init();
    bs_initialized = 1;
  }
  license = StringValuePtr(lic_string);
  return Qtrue;
}

/*
 * BounceStudio#check(raw)
 *  wraps C bsBounceCheck()
 */
static VALUE rb_bs_check(VALUE self, VALUE raw) {
  int BounceCode;
  char *RawMessage, *EmailAddress;

  RawMessage = StringValuePtr(raw);
  
  BounceCode = bsBounceCheck(RawMessage, &EmailAddress, "", license);

  return INT2NUM(BounceCode);
}

/*
 * BounceStudio#get_body(raw)
 *  wraps C bsGetBody()
 */
static VALUE rb_bs_body(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetBody(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#from_address(raw)
 *  wraps C bsGetFromAddress()
 */
static VALUE rb_bs_from_address(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetFromAddress(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#from_friendly_name(raw)
 *  wraps C bsGetFromFriendlyName()
 */
static VALUE rb_bs_from_friendly_name(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetFromFriendlyName(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#header(raw)
 *  wraps C bsGetHeader()
 */
static VALUE rb_bs_header(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetHeader(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#reply_to_address(raw)
 *  wraps C bsGetReplyToAddress()
 */
static VALUE rb_bs_reply_to_address(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetReplyToAddress(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#reply_to_friendly_name(raw)
 *  wraps C bsGetReplyToFriendlyName()
 */
static VALUE rb_bs_reply_to_friendly_name(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetReplyToFriendlyName(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#subject(raw)
 *  wraps C bsGetSubject()
 */
static VALUE rb_bs_subject(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetSubject(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#to_address(raw)
 *  wraps C bsGetToAddress()
 */
static VALUE rb_bs_to_address(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetToAddress(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#to_friendly_name(raw)
 *  wraps C bsGetToFriendlyName()
 */
static VALUE rb_bs_to_friendly_name(VALUE self, VALUE raw) {
  char *RawMessage;
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  bsGetToFriendlyName(RawMessage, &ReturnValue);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#custom_header(raw)
 *  wraps C bsGetCustomHeader()
 */
static VALUE rb_bs_custom_header(VALUE self, VALUE raw, VALUE header) {
  char *RawMessage;
  char *Header; 
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  Header = StringValuePtr(header);
  bsGetCustomHeader(RawMessage, &ReturnValue, Header);

  return rb_str_new2(ReturnValue);
}

/*
 * BounceStudio#orig_custom_header(raw)
 *  wraps C bsGetOrigCustomHeader()
 */
static VALUE rb_bs_orig_custom_header(VALUE self, VALUE raw, VALUE header) {
  char *RawMessage;
  char *Header; 
  char *ReturnValue = NULL; 

  RawMessage = StringValuePtr(raw);
  Header = StringValuePtr(header);
  bsGetOrigCustomHeader(RawMessage, &ReturnValue, Header);

  return rb_str_new2(ReturnValue);
}

/*
 * Extention initializer
 */
void Init_bounce_studio() {
  VALUE boogie_tools_mod =
    rb_define_module("BoogieTools");

  VALUE bs_class =
    rb_define_class_under(boogie_tools_mod,
    "BounceStudio", rb_cObject);

  rb_define_method(bs_class, "initialize", rb_bs_init, 1);
  rb_define_method(bs_class, "check", rb_bs_check, 1);
  rb_define_method(bs_class, "body", rb_bs_body, 1);
  rb_define_method(bs_class, "from_address", rb_bs_from_address, 1);
  rb_define_method(bs_class, "from_friendly_name", rb_bs_from_friendly_name, 1);
  rb_define_method(bs_class, "header", rb_bs_header, 1);
  rb_define_method(bs_class, "reply_to_address", rb_bs_reply_to_address, 1);
  rb_define_method(bs_class, "reply_to_friendly_name", rb_bs_reply_to_friendly_name, 1);
  rb_define_method(bs_class, "subject", rb_bs_subject, 1);
  rb_define_method(bs_class, "to_address", rb_bs_to_address, 1);
  rb_define_method(bs_class, "to_friendly_name", rb_bs_to_friendly_name, 1);
  rb_define_method(bs_class, "custom_header", rb_bs_custom_header, 2);
  rb_define_method(bs_class, "orig_custom_header", rb_bs_orig_custom_header, 2);
}

